### ISSUE TITLE
MySQLSyntaxErrorException: Duplicate key name 'entityid'

### ISSUE THREAD

**Issue by Phoenix616** [LogBlock/LogBlock#815]
Build 114 throws the following error on startup ~~when updating from a previous build (112)~~ every time actually:

```
[13:40:08 ERROR]: [LogBlock] [Updater] Error:
com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: Duplicate key name 'entityid'
        at jdk.internal.reflect.GeneratedConstructorAccessor66.newInstance(Unknown Source) ~[?:?]
        at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
        at java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500) ~[?:?]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:481) ~[?:?]
        at com.mysql.jdbc.Util.handleNewInstance(Util.java:403) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at com.mysql.jdbc.Util.getInstance(Util.java:386) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:944) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3933) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3869) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at com.mysql.jdbc.MysqlIO.sendCommand(MysqlIO.java:2524) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at com.mysql.jdbc.MysqlIO.sqlQueryDirect(MysqlIO.java:2675) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2465) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2439) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at com.mysql.jdbc.StatementImpl.executeInternal(StatementImpl.java:829) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at com.mysql.jdbc.StatementImpl.execute(StatementImpl.java:729) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at de.diddiz.lib.com.zaxxer.hikari.pool.ProxyStatement.execute(ProxyStatement.java:95) ~[?:?]
        at de.diddiz.lib.com.zaxxer.hikari.pool.HikariProxyStatement.execute(HikariProxyStatement.java) ~[?:?]
        at de.diddiz.LogBlock.Updater.update(Updater.java:738) ~[?:?]
        at de.diddiz.LogBlock.LogBlock.onEnable(LogBlock.java:96) ~[?:?]
        at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:263) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at org.bukkit.plugin.java.JavaPluginLoader.enablePlugin(JavaPluginLoader.java:380) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:557) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at org.bukkit.craftbukkit.v1_16_R3.CraftServer.enablePlugin(CraftServer.java:501) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at org.bukkit.craftbukkit.v1_16_R3.CraftServer.enablePlugins(CraftServer.java:415) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at net.minecraft.server.v1_16_R3.MinecraftServer.loadWorld(MinecraftServer.java:465) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at net.minecraft.server.v1_16_R3.DedicatedServer.init(DedicatedServer.java:240) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at net.minecraft.server.v1_16_R3.MinecraftServer.w(MinecraftServer.java:936) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at net.minecraft.server.v1_16_R3.MinecraftServer.lambda$a$0(MinecraftServer.java:174) ~[craftbukkit.jar:git-Origami-"de51baa"]
        at java.lang.Thread.run(Thread.java:832) [?:?]
```

As far as I can tell the update statement could use `ADD KEY IF NOT EXISTS entityid (entityid)` to avoid that.

**Comment by Brokkonaut** [comment]
As far as I know there is not ADD KEY IF NOT EXISTS syntax in mysql.

Logblock just tries to add that key (because it should not be there in not upgraded databases) and if it cannot add it, it should just log that exception but continue normally. And it should update the config version, to just do those table modifications one time.
https://github.com/LogBlock/LogBlock/blob/81e013504680e031d4d1a34d91d0ff5c1bc49134/src/main/java/de/diddiz/LogBlock/Updater.java#L732-L747
I tested this by manually reverting the version in the config file and it worked as expected.

There might be another issue that logblock reaches the connection limit, because connections were not closed properly. Can you please try build 115, it should fix this issue (and it should fix #816 too)

**Comment by Phoenix616** [comment]
> As far as I know there is not ADD KEY IF NOT EXISTS syntax in mysql.
Oh, you are right, it's a MariaDB addition. Do people still use MySQL? Pretty sure mysql has been aliases to MariaDB on most distros nowadays. Might be worth including that info in metrics to see if it is possible to move to the more modern SQL server...

> Logblock just tries to add that key (because it should not be there in not upgraded databases) and if it cannot add it, it should just log that exception but continue normally. And it should update the config version, to just do those table modifications one time.
> https://github.com/LogBlock/LogBlock/blob/81e013504680e031d4d1a34d91d0ff5c1bc49134/src/main/java/de/diddiz/LogBlock/Updater.java#L732-L747
> 
> I tested this by manually reverting the version in the config file and it worked as expected.

Yeah we run an unconventional setup where multiple servers access the same log database for... reasons. So that would've probably been it.

> There might be another issue that logblock reaches the connection limit, because connections were not closed properly. Can you please try build 115, it should fix this issue (and it should fix #816 too)

Yeah that build fixes that situation it and makes some sense. Why not use a try-with-statement for automatic connection closing there?

**Comment by Brokkonaut** [comment]
> Oh, you are right, it's a MariaDB addition. Do people still use MySQL? Pretty sure mysql has been aliases to MariaDB on most distros nowadays. Might be worth including that info in metrics to see if it is possible to move to the more modern SQL server...

I don't know, but as long as keeping mysql compatibility has no major drawbacks, I prefer not to require mariadb.

> Why not use a try-with-statement for automatic connection closing there?

Uhm.. I learned the verbose syntax and it's not easy to abandon a habit ;)
